src/lib.rs

```diff
@@ -1,334 +1,201 @@
+mod protobuf_impl;
 
 #[cfg(feature = "prost-codec")]
-mod wrapper;
+mod prost_impl;
 
-/// Generate Rust files from proto files (`files`).
-pub fn generate_files<T: AsRef<Path> + Debug>(includes: &[T], files: &[T], out_dir: &str) {
-    if Path::new(&out_dir).exists() {
-        fs::remove_dir_all(&out_dir).unwrap();
-    }
-    fs::create_dir_all(&out_dir).unwrap();
-    generate(includes, files, out_dir);
-    let mut generated = Vec::new();
-    let modules: Vec<_> = fs::read_dir(out_dir)
-        .unwrap()
-        .filter_map(|res| {
-            let path = match res {
-                Ok(e) => e.path(),
-                Err(e) => panic!("failed to list {}: {:?}", out_dir, e),
-            };
-            if path.extension() == Some(std::ffi::OsStr::new("rs")) {
-                generated.push(format!("{}", path.display()));
-                let name = path.file_stem().unwrap().to_str().unwrap();
-                Some((name.replace('-', "_"), name.to_owned()))
-            } else {
-                None
-            }
-        })
-        .collect();
-    let mut f = File::create(format!("{}/mod.rs", out_dir)).unwrap();
-    for (module, file_name) in &modules {
-        if cfg!(feature = "protobuf-codec") {
-            writeln!(f, "pub mod {};", module).unwrap();
-            continue;
-        }
-        if module.starts_with("wrapper_") {
-            continue;
-        }
-        let mut level = 0;
-        for part in module.split('.') {
-            writeln!(f, "{:level$}pub mod {} {{", "", part, level = level).unwrap();
-            level += 1;
-        }
-        writeln!(
-            f,
-            "{:level$}include!(\"{}.rs\");",
-            "",
-            file_name,
-            level = level
-        )
-        .unwrap();
-        if Path::new(&format!("{}/wrapper_{}.rs", out_dir, file_name)).exists() {
-            writeln!(
-                f,
-                "{:level$}include!(\"wrapper_{}.rs\");",
-                "",
-                file_name,
-                level = level
-            )
-            .unwrap();
-        }
-        for l in (0..level).rev() {
-            writeln!(f, "{:1$}}}", "", l).unwrap();
-        }
-    }
+use bitflags::bitflags;
+use std::fs::{self, File};
+use std::io::Write;
+use std::path::{Path, PathBuf};
+
+pub struct Builder {
+    files: Vec<String>,
+    includes: Vec<String>,
+    include_black_list: Vec<String>,
+    out_dir: String,
+    wrapper_opts: GenOpt,
 }
 
-/// Use rust-protobuf to generate Rust files from proto files (`files`).
-#[cfg(feature = "protobuf-codec")]
-mod protobuf_imps {
-    use regex::Regex;
-    use std::env;
-    use std::fmt::Debug;
-    use std::fs::{self, File};
-    use std::io::{Read, Write};
-    use std::path::Path;
-    use std::process::Command;
-    use std::str::from_utf8;
+impl Builder {
+    pub fn new() -> Builder {
+        Builder {
+            files: Vec::new(),
+            includes: vec!["include".to_owned(), "proto".to_owned()],
```

Comment:
I think it's bad to assume those directories. It add extras restrictions on how project layouts.

Reply:
This is what we want in all our cases and they can be easily overridden, so it seems nice to have them as the defaults.